net: sockets: Add locking to receive callback

jukkar opened the pull request:
Fix a regression when application is waiting data but does
not notice that because socket layer is not woken up.

This could happen because application was waiting condition
variable but the signal to wake the condvar came before the
wait started. Normally if there is constant flow of incoming
data to the socket, the signal would be given later. But if
the peer is waiting that Zephyr replies, there might be a
timeout at peer.

The solution is to add locking in socket receive callback so
that we only signal the condition variable after we have made
sure that the condition variable is actually waiting the data.

Fixes #34964

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: sockets: Add locking to receive callback

Fix a regression when application is waiting data but does
not notice that because socket layer is not woken up.

This could happen because application was waiting condition
variable but the signal to wake the condvar came before the
wait started. Normally if there is constant flow of incoming
data to the socket, the signal would be given later. But if
the peer is waiting that Zephyr replies, there might be a
timeout at peer.

The solution is to add locking in socket receive callback so
that we only signal the condition variable after we have made
sure that the condition variable is actually waiting the data.

Fixes #34964

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@pfalcon I did testing with `dumb_http_server` and ab and did not see the error any more. Can you give it a run too as you seemed to have a good test setup for this.

jukkar committed:
net: sockets: Release the socket lock if needed

If we are waiting all the data i.e., the MSG_WAITALL flag is set,
then if we have not yet received all the data at the end of the
receive loop. We must use the condition variable to get the signal
when the data is ready to be received. Otherwise the receive loop
will not release the socket lock and receive_cb will not be able
to indicate that data is received.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1221,6 +1221,16 @@ static inline ssize_t zsock_recv_stream(struct net_context *ctx,
 			}
 		}
+
+		if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT) && waitall &&
```
The idea seems fine, but I wonder if the logic could be refined a bit to make it less complex. Currently we call `wait_data()` before entering the loop and at the end of the loop. I think it'd be more reasonable to call the `wait_data()` at the beginning of the loop instead, i. e.  add
```
if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT) {
	res = wait_data(ctx, &timeout);
	if (res < 0) { 
		errno = -res; 
		return -1; 
	}
}
```
after `if (sock_is_eof(ctx))` check at line 1147.

jukkar replied:
I tried that, but it did not work (unit tests started to fail). I can investigate more what was the reason.

jukkar replied:
> I tried that, but it did not work (unit tests started to fail). I can investigate more what was the reason.

My earlier experiment with this was clearly wrong as now, when trying it again, it works just fine. Will send a new version shortly.

jukkar committed:
net: sockets: Do not hijack k_fifo API name

The k_fifo_ prefix is meant for kernel API functions, and
not to our socket helper. So remove the k_ prefix in order
to avoid confusion.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
@jukkar 

> I did testing with dumb_http_server and ab and did not see the error any more. Can you give it a run too as you seemed to have a good test setup for this.

Sorry for the delay with response, I was on vacation. Confirming that that this patch resolved issue with dumb_http_server/qemu_x86 + ab for me. Also retested with big_http_download (as the test for "opposite" transfer direction), and it works as expected.

Thanks for investigating this!
